File: server/scraper/websites/servisSta.ts
import { INews } from '../../model/News';
import cheerio from 'cheerio';
import axios from 'axios';
import { parse } from 'date-fns';
import * as Db from '../../db/db';

function extractCountryFromTitle(title: string): string {
  const countryPrefixes = ['v ', 'v ', 'na '];

  for (const prefix of countryPrefixes) {
    const startIndex = title.indexOf(prefix);
    if (startIndex !== -1) {
      const endIndex = title.indexOf(' ', startIndex + prefix.length);
      if (endIndex !== -1) {
        return title.substring(startIndex + prefix.length, endIndex);
      }
    }
  }

  return '';
}

async function servisSta(numArticlesToOpen: number): Promise<INews[]> {
  const response = await axios.get('https://servis.sta.si/');
  const $ = cheerio.load(response.data);

  const newsList: INews[] = [];

  const articleElements = $('article.item');

  for (let i = 0; i < numArticlesToOpen && i < articleElements.length; i++) {
    const articleElement = articleElements.eq(i);

    const url = articleElement.find('a').attr('href');

    if (!url) {
      continue; // Skip this iteration if url is undefined
    }

    const articleResponse = await axios.get(`https://servis.sta.si${url}`);
    const article$ = cheerio.load(articleResponse.data);

    const title = article$('article.articleui h1')
      .text()
      .replace(/\n/g, '')
      .trim();

    const leadElement = article$('article').find('.lead');
    const lead = leadElement.text() || '';

    const textElements = article$('article').find('.text');
    const preTextElement = textElements.eq(0).find('pre');
    const preText = preTextElement.length > 0 ? preTextElement.text() : '';

    const contentTmp = lead || preText ? `${lead} ${preText}` : '';
    const content = contentTmp.replace(/\n/g, '').trim();

    const categoryElement = article$('aside.articlemeta').find(
      'div.items > div:nth-child(2)'
    );
    const categories = [
      categoryElement.text().replace('Kategorija:', '').trim().toLowerCase(),
    ];

    const authorElement = article$('aside.articlemeta').find(
      'div.items > div:nth-child(4)'
    );
    const authors = authorElement
      .text()
      .replace('Avtor:', '')
      .trim()
      .split('/');

    let location = extractCountryFromTitle(title);
    if(location == '') {
        location = Db.Util.getFirstSettlement(categories);
    }

    const coords = await Db.Util.toCoords(location);

    const date = parse('2000-10-10', 'yyyy-MM-dd', new Date());

    const news: INews = {
      title,
      url: `https://servis.sta.si${url}`,
      date,
      authors,
      content,
      categories,
      views: [],
      location: {
        type: 'Point',
        coordinates: coords,
      },
    };

    newsList.push(news);

    // Wait or introduce delay if necessary before making the next request
  }

  return newsList;
}

export = servisSta;

File: desktop-app/src/jvmMain/kotlin/scraper/servisSta.kt
import org.example.model.INews
import org.example.model.Location
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.chrome.ChromeOptions
import java.text.SimpleDateFormat
import java.util.Date

fun extractCountryFromTitle(title: String): String {
    val countryPrefixes = listOf("v ", "v ", "na ")

    for (prefix in countryPrefixes) {
        val startIndex = title.indexOf(prefix)
        if (startIndex != -1) {
            val endIndex = title.indexOf(" ", startIndex + prefix.length)
            if (endIndex != -1) {
                return title.substring(startIndex + prefix.length, endIndex)
            }
        }
    }

    return ""
}

fun getServisSta(numArticlesToOpen: Int): List<INews> {

    val options = ChromeOptions()
    options.addArguments("--headless=new");
    val driver: WebDriver = ChromeDriver(options)

    driver.get("https://servis.sta.si/")


    val newsList = mutableListOf<INews>()

    for (i in 0 until numArticlesToOpen) {
        val articles: List<WebElement> = driver.findElements(By.className("item"))

        val article: WebElement = articles[i]

        val url: String = article.findElement(By.cssSelector("a")).getAttribute("href")
        driver.get(url)

        val title = driver.findElement(By.cssSelector("article.articleui h1")).text

        val articleElement = driver.findElement(By.tagName("article"))
        val leadElement = try {
            articleElement.findElement(By.className("lead"))
        } catch (e: NoSuchElementException) {
            null
        }
        val lead = leadElement?.text ?: ""

        val textElements = articleElement.findElements(By.className("text"))
        val preText = if (textElements.isNotEmpty()) {
            val preTextElement = textElements[0]
            val preElements = preTextElement.findElements(By.tagName("pre"))
            if (preElements.isNotEmpty()) {
                preElements[0].text
            } else {
                ""
            }
        } else {
            ""
        }

        val content = if (lead.isNotEmpty() && preText.isNotEmpty()) {
            "$lead $preText"
        } else if (lead.isNotEmpty()) {
            lead
        } else if (preText.isNotEmpty()) {
            preText
        } else {
            ""
        }

        val categoryElement = driver.findElement(By.cssSelector("aside.articlemeta div.items > div:nth-child(2)"))
        val categories = listOf(categoryElement.text.replace("Kategorija:", "").trim())

        val authorElement = driver.findElement(By.cssSelector("aside.articlemeta div.items > div:nth-child(4)"))
        val authors = authorElement.text.replace("Avtor:", "").trim().split("/")


        val location = extractCountryFromTitle(title)

        val date: Date = SimpleDateFormat("yyyy-MM-dd").parse("2000-10-10")

        val news = INews(title, url, date, authors, content, categories, Location(
            type = "Point",
            coordinates = Pair(0.0,0.0),
        )
        )
        newsList.add(news)

        driver.navigate().back()
    }

    driver.quit()

    return newsList
}
